lib/trie/database.go

```diff
@@ -61,241 +61,267 @@ func (t *Trie) store(db chaindb.Batch, curr Node) error {
 // It recursively descends into the trie using the database starting
 // from the root node until it reaches the node with the given key.
 // It then reads the value from the database.
-func GetFromDB(db chaindb.Database, root common.Hash, key []byte) ([]byte, error) {
-	if root == EmptyHash {
+func GetFromDB(db chaindb.Database, rootHash common.Hash, key []byte) (
+	value []byte, err error) {
+	if rootHash == EmptyHash {
 		return nil, nil
 	}
 
 	k := codec.KeyLEToNibbles(key)
 
-	enc, err := db.Get(root[:])
+	encodedRootNode, err := db.Get(rootHash[:])
 	if err != nil {
-		return nil, fmt.Errorf("failed to find root key=%s: %w", root, err)
+		return nil, fmt.Errorf("cannot find root hash key 0x%x: %w", rootHash, err)
 	}
 
-	rootNode, err := node.Decode(bytes.NewReader(enc))
+	reader := bytes.NewReader(encodedRootNode)
+	rootNode, err := node.Decode(reader)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("cannot decode root node: %w", err)
 	}
 
 	return getFromDB(db, rootNode, k)
 }
 
-func getFromDB(db chaindb.Database, parent Node, key []byte) ([]byte, error) {
-	var value []byte
-
-	switch p := parent.(type) {
-	case *node.Branch:
-		length := lenCommonPrefix(p.Key, key)
-
-		// found the value at this node
-		if bytes.Equal(p.Key, key) || len(key) == 0 {
-			return p.Value, nil
-		}
-
-		// did not find value
-		if bytes.Equal(p.Key[:length], key) && len(key) < len(p.Key) {
-			return nil, nil
-		}
-
-		if p.Children[key[length]] == nil {
-			return nil, nil
+// getFromDB recursively searches through the trie and database
+// for the value corresponding to a key.
+// Note it does not copy the value so modifying the value bytes
+// slice will modify the value of the node in the trie.
+func getFromDB(db chaindb.Database, n Node, key []byte) (
+	value []byte, err error) {
+	leaf, ok := n.(*node.Leaf)
+	if ok {
+		if bytes.Equal(leaf.Key, key) {
+			return leaf.Value, nil
 		}
+		return nil, nil
+	}
 
-		// load child with potential value
-		enc, err := db.Get(p.Children[key[length]].GetHash())
-		if err != nil {
-			return nil, fmt.Errorf("failed to find node in database: %w", err)
-		}
+	branch := n.(*node.Branch)
+	// Key is equal to the key of this branch or is empty
+	if len(key) == 0 || bytes.Equal(branch.Key, key) {
+		return branch.Value, nil
+	}
 
-		child, err := node.Decode(bytes.NewReader(enc))
-		if err != nil {
-			return nil, err
-		}
+	commonPrefixLength := lenCommonPrefix(branch.Key, key)
+	if len(key) < len(branch.Key) && bytes.Equal(branch.Key[:commonPrefixLength], key) {
+		// The key to search is a prefix of the node key and is smaller than the node key.
+		// Example: key to search: 0xabcd
+		//          branch key:    0xabcdef
+		return nil, nil
+	}
 
-		value, err = getFromDB(db, child, key[length+1:])
-		if err != nil {
-			return nil, err
-		}
-	case *node.Leaf:
-		if bytes.Equal(p.Key, key) {
-			return p.Value, nil
-		}
-	case nil:
```

Comment:
I am guessing this would have never happened, because go does not permit use of untyped nil `use of untyped nil`. parent has to be one of node or leaf.

Reply:
It can happen if the parent is `nil`, for example with an empty trie the root can be `nil` (untyped).